documentation/writing_tests.md

```diff
@@ -0,0 +1,1534 @@
+  });
+
+  afterEach(function () {
+    recorder.stop();
+  });
+
+  it("should test A #node", function() {
+    if (!isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test B #browser", function() {
+    if (isNode) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+
+  it("should test C #live", function() {
+    if (env.TEST_MODE) {
+      return this.skip();
+    }
+    // Test contents..
+  });
+});
+```
+
+In case conditionals might appear to be necessary in other scenarios, first consider separating the test case into as many test cases as necessary to remove the conditionals.
+
+You can see a working example of conditionals in our [conditionals.spec.ts test in our template project](../sdk/template/template/test/public/conditionals.spec.ts).
+
+### Using delays
+
+The API methods that we provide should always allow users to wait until the service operation finishes. For this reason, any method that doesn't immediately respond with a completed operation should use our [Long Running Operation strategy (core-lro)](https://github.com/Azure/azure-sdk-for-js/tree/master/sdk/core/core-lro).
+
+If a method can't use `core-lro` (for example, if the service cannot be trusted to provide signs of completion, or finish within reasonable time), and thus we see the need to use a different delay strategy, we should always wait until the next possible operation can be fulfilled. **Tests that need to wait for an operation that cannot be trusted to finish reasonably should not run during the automated execution of live tests**. The following snippet can be used to skip these tests in live mode:
+
+```ts
+  // `isRecordMode` and `isPlaybackMode` are methods exported from the Recorder.
+  if (!isRecordMode() && !isPlaybackMode()) {
+    return this.skip();
+  }
+```
+
+For example, in `@azure/keyvault-keys` we provide a method to purge keys that has not been moved to use `core-lro`. To check that it has finished, we do a while loop where we try to make the next operation until it passes:
+
+```ts
+import { delay, isRecordMode, isPlaybackMode } from "@azure/test-utils-recorder";
+// ...
+
+describe("Keys client - restore keys and recover backups", () => {
+  // ...
+  it("can restore a key with a given backup", async function() {
+
+    // This test can't be expected to finish in any reasonable time.
+    if (!isRecordMode() && !isPlaybackMode()) {
+      return this.skip();
+    }
+
+    // ...
+    await client.createKey(keyName, "RSA");
+    const backup = await client.backupKey(keyName);
+    await client.purgeDeletedKey(keyName);
+    while (true) {
+      try {
+        await client.restoreKeyBackup(backup as Uint8Array);
+        break;
+      } catch (e) {
+        console.log("Can't restore the key since it's not fully deleted:", e.message);
+        console.log("Retrying in one second...");
+        // This delay method comes from the recorder
+        await delay(1000);
```

Comment:
seems like this delay number should be its own const

Reply:
Interesting! Why?

Reply:
oh I think this is just me having the general sense that random number literals in code are to be avoided because it's not clear what they mean or if they get used in multiple places it's hard to update.

e.g.

```typescript
const KEY_DELETION_DELAY_IN_MS = 1000;
```

but I don't feel too strongly about this since it's the only usage of the literal.

Reply:
Pushing this!